pkg/bicep/bicep.go

```diff
@@ -23,19 +29,109 @@ var template = map[string]interface{}{
 }
 
 func GenerateFiles(buildPath, stepPath string, b *bundle.Bundle, fs afero.Fs) error {
-	p := path.Join(buildPath, stepPath, paramsFile)
+	p := path.Join(buildPath, stepPath, ParamsFile)
 	params, _ := inputvars.TranspileDevParams(p, b, fs)
 	connections, _ := inputvars.TranspileConnectionVarFile(p, b, fs)
 
-	template["parameters"].(map[string]interface{})["params"] = params
-	template["parameters"].(map[string]interface{})["md_metadata"] = inputvars.DefaultMdMetadata(b.Name)
-	template["parameters"].(map[string]interface{})["connections"] = connections
+	template["parameters"].(map[string]interface{})["params"] = map[string]any{"value": params}
+	template["parameters"].(map[string]interface{})["md_metadata"] = map[string]any{"value": inputvars.DefaultMdMetadata(b.Name)}
+	template["parameters"].(map[string]interface{})["connections"] = map[string]any{"value": connections}
 
 	content, err := json.MarshalIndent(template, "", "    ")
 
 	if err != nil {
 		return err
 	}
 
-	return afero.WriteFile(fs, fmt.Sprintf("%s/%s", stepPath, paramsFile), content, 0755)
+	return afero.WriteFile(fs, fmt.Sprintf("%s/%s", stepPath, ParamsFile), content, 0755)
+}
+
+// reconcileParams reads the params file keeping the md_metadata field intact as it's
+// not represented in the UI yet, adds the incoming params, and writes the file back out.
+func ReconcileParams(baseDir string, params map[string]any) error {
+	paramPath := path.Join(baseDir, ParamsFile)
+
+	fileParams := make(map[string]any)
+	err := files.Read(paramPath, &fileParams)
+	if err != nil {
+		return err
+	}
+
+	fileParams["parameters"].(map[string]interface{})["params"] = params
+
+	return files.Write(paramPath, fileParams)
+}
+
+func NewInputHandler() InputHandler {
+	return InputHandler{
+		ParamsFile:      ParamsFile,
+		ConnectionsFile: ParamsFile,
+	}
+}
+
+func (i InputHandler) ReadParams(basePath string) ([]byte, error) {
```

Comment:
How does this ReadParams work with existing bicep templates? Do they need to restructure their template to adhere to the params/connections hierarchy, and filename?

I ask because terraform work is in a PR that is able to read variables (regardless of file location), and only render new variables that are declared in `massdriver.yaml` but not in the terraform module. Whether they are a param/connection is arbitrary (they can be determined if necessary by looking at the variable name).

Reply:
It is definetly not optimized for that stuff. I don't know enough bicep to know how people are using it. The way the type system works in bicep is sort of tricky so this is doing the absolute least. We can hone it in later. right now you can make outputs by defining three lines of any in a bicep template and rolling.

Reply:
If I understand you correctly, I think at the very least we should probably flatten the inputs to remove the top level `params` and `connections`. This makes it so all existing bicep templates are directly importable (when we have time to implement that). Otherwise existing bicep templates will need to be changed to fit the "massdriver" standard of having a top-level `params` and `connections` key.

We already have linting checks to ensure that params and connections don't share keys, so they can be at the same level. This is how it works w/ terraform and helm. We don't have a top level `param` variable (or helm value) with all the params under, and `connections` variable (or helm value) with all connections under it. They are all just top level variables/values.

But if I'm not understanding correctly, please clarify.

Reply:
Sounds good. I don't know enough about how people manage bicep so I just did the easiest thing to explain in a demo.

Reply:
I can make that change though